<h1><center>Laboratorio 6: Optimización de modelos 🧪</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2025</strong></center>

### Cuerpo Docente:

- Profesores: Diego Cortez, Gabriel Iturra
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Nicolás Cabello, Cristopher Urbina

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Josefa Anselmo.
- Nombre de alumno 2: Tamara Carrasco.


Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: Entregas Martes a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria, pero se recomienda **fuertemente** asistir.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### **Link de repositorio de GitHub:** [TamoJosha](https://github.com/Tamaracarrasco/TamoJosha-Lab-de-Prog-MDS-Prim2025)

### Temas a tratar

- Predicción de demanda usando `xgboost`
- Búsqueda del modelo óptimo de clasificación usando `optuna`
- Uso de pipelines.


### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Martes a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria, pero se recomienda fuertemente asistir.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# El emprendimiento de Fiu

Tras liderar de manera exitosa la implementación de un proyecto de ciencia de datos para caracterizar los datos generados en Santiago 2023, el misterioso corpóreo **Fiu** se anima y decide levantar su propio negocio de consultoría en machine learning. Tras varias e intensas negociaciones, Fiu logra encontrar su *primera chamba*: predecir la demanda (cantidad de venta) de una famosa productora de bebidas de calibre mundial. Al ver el gran potencial y talento que usted ha demostrado en el campo de la ciencia de datos, Fiu lo contrata como data scientist para que forme parte de su nuevo emprendimiento.

Para este laboratorio deben trabajar con los datos `sales.csv` subidos a u-cursos, el cual contiene una muestra de ventas de la empresa para diferentes productos en un determinado tiempo.

Para comenzar, cargue el dataset señalado y visualice a través de un `.head` los atributos que posee el dataset.

<i><p align="center">Fiu siendo felicitado por su excelente desempeño en el proyecto de caracterización de datos</p></i>
<p align="center">
  <img src="https://media-front.elmostrador.cl/2023/09/A_UNO_1506411_2440e.jpg">
</p>

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime
!pip install -qq xgboost optuna

df = pd.read_csv("sales.csv")

df.head(10)

id      date    city       lat      long     pop    shop         brand  \
0   0  31/01/12  Athens  37.97945  23.71622  672130  shop_1   kinder-cola   
1   1  31/01/12  Athens  37.97945  23.71622  672130  shop_1   kinder-cola   
2   2  31/01/12  Athens  37.97945  23.71622  672130  shop_1   kinder-cola   
3   3  31/01/12  Athens  37.97945  23.71622  672130  shop_1    adult-cola   
4   4  31/01/12  Athens  37.97945  23.71622  672130  shop_1    adult-cola   
5   5  31/01/12  Athens  37.97945  23.71622  672130  shop_1  orange-power   
6   6  31/01/12  Athens  37.97945  23.71622  672130  shop_1  orange-power   
7   7  31/01/12  Athens  37.97945  23.71622  672130  shop_1        gazoza   
8   8  31/01/12  Athens  37.97945  23.71622  672130  shop_1   lemon-boost   
9   9  31/01/12  Athens  37.97945  23.71622  672130  shop_1   lemon-boost   

  container capacity  price  quantity  
0     glass    500ml   0.96     13280  
1   plastic    1.5lt   2.86      6727  
2       can    330ml   0.87      9848  
3     glass    500ml   1.00     20050  
4       can    330ml   0.39     25696  
5     glass    500ml   1.00     15041  
6       can    330ml   0.43     34578  
7     glass    500ml   0.49     44734  
8     glass    500ml   0.70     18623  
9   plastic    1.5lt   2.21      9645

In [2]:
## Exploración básica de los datos

print("\nDimensiones del dataset:\n", df.shape)
print("\n Tipo de valores en las columnas: \n", df.info())
print("\n Cantidad de valores nulos: \n", df.isna().sum())
print("\n Valores duplicados: \n", df.duplicated().sum())
print("\n describe de las variables númericas: \n", df.select_dtypes(include="number").describe())
print("\n Estadísticas de las variables no numéricas: \n", df.select_dtypes(include="object").describe())

## Rango temporal de los datos.

print("\n ========= RANGO TEMPORAL (Aparente) =========")
print("Fecha inicial: ", df["date"].min())
print("última fecha: ", df["date"].max())


Dimensiones del dataset:
 (7456, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7456 entries, 0 to 7455
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         7456 non-null   int64  
 1   date       7456 non-null   object 
 2   city       7456 non-null   object 
 3   lat        7456 non-null   float64
 4   long       7456 non-null   float64
 5   pop        7456 non-null   int64  
 6   shop       7456 non-null   object 
 7   brand      7456 non-null   object 
 8   container  7456 non-null   object 
 9   capacity   7456 non-null   object 
 10  price      7456 non-null   float64
 11  quantity   7456 non-null   int64  
dtypes: float64(3), int64(3), object(6)
memory usage: 699.1+ KB

 Tipo de valores en las columnas: 
 None

 Cantidad de valores nulos: 
 id           0
date         0
city         0
lat          0
long         0
pop          0
shop         0
brand        0
container    0
capacity     0
price   

### **Comentarios generales de la base**

De las estadísticas y exploraciones iniciales se desprende lo siguiente:

- **No existen valores nulos** en ninguna columna.  
- **No hay valores duplicados**, lo que indica una correcta unicidad por registro.  
- **No se observan valores anómalos evidentes** (por ejemplo, cantidades negativas en `quantity` o precios fuera de rango lógico).  
- La variable **`quantity`** (target) es continua, por lo tanto, el problema corresponde a una **tarea de regresión supervisada**.  
- Dentro de las variables categóricas, **`capacity`** es una variable **categórica ordinal**, ya que sus niveles representan distintos volúmenes ordenables (*330ml < 500ml < 1.5lt*).  
- El resto de variables categóricas (`city`, `shop`, `brand`, `container`) son **nominales**.  
- **`date`** representa una **serie temporal discreta** que abarca el período **2013–2018**, lo que permite derivar variables de estacionalidad tales como *año*, *mes* o *día de la semana*, útiles para capturar patrones de demanda a lo largo del tiempo.

---

#### **Distribución geográfica y estructura de tiendas**
- Se identifican **5 ciudades**: `Athens`, `Thessaloniki`, `Patra`, `Larisa`, `Irakleion`.
- Se registran **6 tiendas** distintas (`shop_1` a `shop_6`), cada una asociada a una única ciudad, lo que sugiere una estructura **jerárquica ciudad–tienda**:
  - `shop_1` → *Athens*  
  - `shop_2` → *Irakleion*  
  - `shop_3` → *Athens*  
  - `shop_4` → *Thessaloniki*  
  - `shop_5` → *Larisa*  
  - `shop_6` → *Patra*  

  Esta relación indica que `city` y `shop` están **altamente correlacionadas** (una relación 1:N), por lo que conviene **evitar incluir ambas sin control de colinealidad**, o bien tratarlas mediante *encoding jerárquico* o eliminación de redundancia.

---

#### **Marcas y envases**
- Se identifican **5 marcas**:
  - `kinder-cola`: *plastic, can, glass*  
  - `adult-cola`: *plastic, can, glass*  
  - `lemon-boost`: *glass, can, plastic*  
  - `gazoza`: *glass, plastic, can*  
  - `orange-power`: *glass, can, plastic*  

- Existen **3 tipos de envase (`container`)**:
  - `glass`: con capacidades de *500ml* y *330ml*.  
  - `can`: solo disponible en *330ml*.  
  - `plastic`: solo disponible en *1.5lt*.  

  Esto revela una **dependencia estructural entre `container` y `capacity`**, por lo que se recomienda **no codificarlas de forma completamente independiente**, ya que parte de la información se solapa.

---

#### **Variables numéricas**
- `lat` y `long` son coordenadas geográficas fijas por ciudad, **fuertemente correlacionadas con `city`**, y podrían eliminarse si se mantiene la variable de ciudad.  
- `pop` (población) varía por ciudad y podría ser relevante como indicador de tamaño de mercado local.  
- `price` presenta un rango entre 0.11 y 4.79, con distribución aparentemente asimétrica (se recomienda inspeccionar outliers).  

---

#### **Variable objetivo**
- La variable **`quantity`** representa la **cantidad vendida** de cada producto en una fecha, tienda y presentación específicas.  
- Dado que se trata de un valor numérico continuo, la tarea corresponde a un **problema de regresión**, en el que el objetivo será **predecir la demanda futura** a partir de atributos de producto, localización, precio y tiempo.  
- Antes de la modelación será necesario **verificar la distribución de `quantity`** (posible sesgo hacia la derecha), y aplicar transformaciones (por ejemplo, logarítmica) si se observan asimetrías significativas que afecten el ajuste del modelo.


## 1 Generando un Baseline (5 puntos)

<p align="center">
  <img src="https://media.tenor.com/O-lan6TkadUAAAAC/what-i-wnna-do-after-a-baseline.gif">
</p>

Antes de entrenar un algoritmo, usted recuerda los apuntes de su magíster en ciencia de datos y recuerda que debe seguir una serie de *buenas prácticas* para entrenar correcta y debidamente su modelo. Después de un par de vueltas, llega a las siguientes tareas:

1. Separe los datos en conjuntos de train (70%), validation (20%) y test (10%). Fije una semilla para controlar la aleatoriedad. [0.5 puntos]
2. Implemente un `FunctionTransformer` para extraer el día, mes y año de la variable `date`. Guarde estas variables en el formato categorical de pandas. [1 punto]
3. Implemente un `ColumnTransformer` para procesar de manera adecuada los datos numéricos y categóricos. Use `OneHotEncoder` para las variables categóricas. `Nota:` Utilice el método `.set_output(transform='pandas')` para obtener un DataFrame como salida del `ColumnTransformer` [1 punto]
4. Guarde los pasos anteriores en un `Pipeline`, dejando como último paso el regresor `DummyRegressor` para generar predicciones en base a promedios. [0.5 punto]
5. Entrene el pipeline anterior y reporte la métrica `mean_absolute_error` sobre los datos de validación. ¿Cómo se interpreta esta métrica para el contexto del negocio? [0.5 puntos]
6. Finalmente, vuelva a entrenar el `Pipeline` pero esta vez usando `XGBRegressor` como modelo **utilizando los parámetros por default**. ¿Cómo cambia el MAE al implementar este algoritmo? ¿Es mejor o peor que el `DummyRegressor`? [1 punto]
7. Guarde ambos modelos en un archivo .pkl (uno cada uno) [0.5 puntos]

In [3]:
# Deja un output y se ve feo que quede en el resultado de la celda que sigue
!pip install autopep8

In [4]:
# importación de librerías importantes.

# separacion de conjuntos
from sklearn.model_selection import train_test_split

# para el pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, FunctionTransformer
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.dummy import DummyRegressor
from xgboost import XGBRegressor

# métricas

from sklearn.metrics import mean_absolute_error as MAE

# para guardar los modelos con pickle
import os
import pickle

from sklearn import set_config
set_config(transform_output="pandas")

# Inserte su código acá

# Separo train - test - val + semilla (Punto 1)

df_sorted = df.copy()
df_sorted["date"] = pd.to_datetime(df_sorted["date"], format="%d/%m/%y", dayfirst=True, errors="coerce")
df_sorted = df_sorted.sort_values("date").reset_index(drop=True)

df_x = df_sorted.drop(columns="quantity")
df_y = df_sorted["quantity"]

X_train, X_val, y_train, y_val = train_test_split(
    df_x, df_y, random_state=42, shuffle=False, train_size=0.7
)
X_val, X_test, y_val, y_test = train_test_split(
    X_val, y_val, test_size=0.33, shuffle=False, random_state=42
)

drop_cols = ["id"]

num_cols = ["lat", "long", "pop", "price"]
cat_cols = ["city", "shop", "brand", "container",
            "capacity", "day", "month", "year"]

# Definición de FunctionTraformer (Punto 2)


def extrae_fecha(dataframe_in, col="date"):
    """
    Esta función debería extraer el día, mes y año
    de la variable "date" de un dataframe.
    Guarda las variables en el formato categórico
    de pandas. Devuelve dataframe sin columna de "date".

    Parámetros:
    -------------------
    dataframe_in: dataframe para extraer los datos. 
    Debe contener columna "date".

    col: columna de la fecha. Default "date".

    """
    # primer transformamos esa columna a una nueva

    df_copy = dataframe_in.copy()

    df_copy["fecha"] = pd.to_datetime(
        df_copy[col], dayfirst=True, errors="coerce", format="%d/%m/%y")

    # se crean las columnas de día, mes y año.

    df_copy["day"] = df_copy["fecha"].dt.day.astype("category")
    df_copy["month"] = df_copy["fecha"].dt.month.astype("category")
    df_copy["year"] = df_copy["fecha"].dt.year.astype("category")

    # Se elimina columna "fecha" y "date"
    df_copy = df_copy.drop(columns=["fecha", "date"]) 

    return df_copy

date_transformer = FunctionTransformer(extrae_fecha, validate=False)

# PIPELINES POR GRUPO (punto 3)

num_std_pipeline = Pipeline([
    ("imp", SimpleImputer(strategy="median")),
    ("sc", StandardScaler())
])

cat_pipeline = Pipeline([
    ("imp", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore", sparse_output=False))
])

col_transformer = ColumnTransformer(
    [
        ("drop_ids", "drop", drop_cols),
        ("numerical", num_std_pipeline, num_cols),
        ("categorical", cat_pipeline, cat_cols)
    ],
    verbose_feature_names_out=False,
    remainder="drop"
).set_output(transform="pandas")


# PIPELINE DEL MODELO (Punto 4)

pipeline_dummyreg = Pipeline(
    steps=[
        ("date_features", date_transformer),
        ("col_transformer", col_transformer),
        ("regresor_dummy", DummyRegressor(strategy="mean"))
    ]
)

# ENTRENAMIENTO DEL MODELO DUMMYREGRESSOR

pipeline_dummyreg.fit(X_train, y_train)

# PREDICCIONES DUMMYREGRESSOR

y_pred_val_dummyreg = pipeline_dummyreg.predict(X_val)

mae_dummy = MAE(y_val, y_pred_val_dummyreg)

#### REPORTE DUMMY MAE para dummyregressor (Punto 5)

print("Valor MAE (DummyRegressor) para conjunto Validación: ", mae_dummy)

### XGBOOST (Punto 6)

pipeline_xgb = Pipeline([
    ("date_features", FunctionTransformer(extrae_fecha, validate=False)),
    ("col_transformer", col_transformer),
    ("regresor_xgb", XGBRegressor(
        objective="reg:squarederror",
        tree_method="hist",
        random_state=42,
        n_jobs=-1
    ))
])

#Entrenamiento y predicciones 
pipeline_xgb.fit(X_train, y_train)
y_pred_val_xgb = pipeline_xgb.predict(X_val)

#Resultados 
mae_xgb = MAE(y_val, y_pred_val_xgb)
print(f"MAE (XGBRegressor) validación: {mae_xgb:,.2f}")
print("ΔMAE vs Dummy:", mae_xgb - mae_dummy)

# Guardado de modelos (Punto 7)

base_path = os.path.join(os.getcwd(), "Laboratorios", "Laboratorio_6", "data")
os.makedirs(base_path, exist_ok=True)

# Predicciones + y_true (validación)
pd.DataFrame({"y_val": y_val, "y_pred_dummy": y_pred_val_dummyreg}).to_csv(
    os.path.join(base_path, "predicciones_dummy_regressor.csv"), index=False
)
pd.DataFrame({"y_val": y_val, "y_pred_xgb": y_pred_val_xgb}).to_csv(
    os.path.join(base_path, "predicciones_xgb_regressor.csv"), index=False
)

# Guardo también el de test de xgboost por sia caso
y_pred_test_xgb = pipeline_xgb.predict(X_test)
pd.DataFrame({"y_test": y_test, "y_pred_xgb": y_pred_test_xgb}).to_csv(
    os.path.join(base_path, "predicciones_xgb_test.csv"), index=False
)

# Guardar modelos
with open(os.path.join(base_path, "dummy_regressor_model.pkl"), "wb") as f:
    pickle.dump(pipeline_dummyreg, f)

with open(os.path.join(base_path, "xgb_regressor_model.pkl"), "wb") as f:
    pickle.dump(pipeline_xgb, f)

# Cargar y probar
with open(os.path.join(base_path, "xgb_regressor_model.pkl"), "rb") as f:
    modelo_loaded = pickle.load(f)
_ = modelo_loaded.predict(X_val[:5])

Valor MAE (DummyRegressor) para conjunto Validación:  12518.387190585083
MAE (XGBRegressor) validación: 2,796.36
ΔMAE vs Dummy: -9722.028059725708


**Respuesta desarrollo punto 5:** El MAE (Mean Absolute Error) obtenido para el conjunto de validación fue de 12 518.39 unidades.
Esta métrica representa el promedio de error absoluto entre las cantidades reales y las predichas, expresado en las mismas unidades de la variable objetivo (cantidad vendida en este caso).
En términos de negocio, significa que, en promedio, el modelo baseline (DummyRegressor) se equivoca en alrededor de 12.500 unidades por observación, por ejemplo, por ciudad, tienda, marca o tipo de contenedor, al estimar la demanda esperada.
Dado que el modelo predice siempre la media histórica, este resultado es esperable y refleja un desempeño sin poder predictivo real, ya que no incorpora variaciones por precio, estacionalidad ni diferencias entre tiendas o formatos.
En consecuencia, el MAE sirve como referencia inicial pues todo modelo posterior deberá superar esta cifra para considerarse útil desde una perspectiva operativa, pues errores de esta magnitud implicarían desbalances significativos en planificación de producción, inventario y distribución.

**Respuesta desarrollo punto 6:** El XGBRegressor (parámetros por defecto) obtuvo un MAE de ~2,800 en validación, frente a 12,518.39 del DummyRegressor. Esto implica una reducción del error absoluto promedio de ~9720 unidades (~78%), por lo que el XGB es claramente mejor que el baseline. La mejora es consistente con un modelo de gradiente boosting porque captura relaciones no lineales e interacción entre variables (precio, ciudad, tienda, marca, envase y estacionalidad) que el Dummy (promedio constante) no modela.

*Cabe mencionar que las cifras son aproximadas porque al correr el modelo localmente en dos equipos, los resultados cambiaban, a pesar de dejar una semilla fija.

## 2. Forzando relaciones entre parámetros con XGBoost (10 puntos)

<p align="center">
  <img src="https://64.media.tumblr.com/14cc45f9610a6ee341a45fd0d68f4dde/20d11b36022bca7b-bf/s640x960/67ab1db12ff73a530f649ac455c000945d99c0d6.gif">
</p>

Un colega aficionado a la economía le *sopla* que la demanda guarda una relación inversa con el precio del producto. Motivado para impresionar al querido corpóreo, se propone hacer uso de esta información para mejorar su modelo realizando las siguientes tareas:

1. Vuelva a entrenar el `Pipeline` con `XGBRegressor`, pero esta vez forzando una relación monótona negativa entre el precio y la cantidad. Para aplicar esta restricción apóyese en la siguiente <a href = https://xgboost.readthedocs.io/en/stable/tutorials/monotonic.html>documentación</a>. [6 puntos]

>Hint 1: Para implementar el constraint se le sugiere hacerlo especificando el nombre de la variable. De ser así, probablemente le sea útil **mantener el formato de pandas** antes del step de entrenamiento.

>Hint 2: Puede obtener el nombre de las columnas en el paso anterior al modelo regresor mediante el método `.get_feature_names_out()`

2. Luego, vuelva a reportar el `MAE` sobre el conjunto de validación. [1 puntos]

3. ¿Cómo cambia el error al incluir esta relación? ¿Tenía razón su amigo? [2 puntos]

4. Guarde su modelo en un archivo .pkl [1 punto]

In [5]:
# Inserte su código acá

# Reentreno o redefino (clonado eso si) el pipe (Punto 1)

# Tengo que obtener nombres de columnas en el orden real del modelo entonces tengo que transformar X_train con los pasos previos al estimador y además, 
# evitar modificar el pipeline ya entrenado usando un clone del ColumnTransformer para que quede menos enredado.

from sklearn.base import clone

# Primero aplico el FunctionTransformer de fecha
X_train_dt = pipeline_xgb.named_steps["date_features"].transform(X_train)

# Ajusto con un clone del col_transformer para evitar problemas después
col_trans_clone = clone(pipeline_xgb.named_steps["col_transformer"]).set_output(transform="pandas")
X_train_trans = col_trans_clone.fit_transform(X_train_dt)

# Saco los nombres de las columnas según el orden que están en el modelo
feature_names = col_trans_clone.get_feature_names_out()
print("Total features:", len(feature_names))

# Inspecciono cuales son y el orden
print("Columnas en orden del df:", feature_names)

# Después toca construir el mapeo de constraints, que lo vamos a definir como 0 para todas las features y -1 para 'price' 
# (como por la relación inversa con quantity)

monotone_map = {name: 0 for name in feature_names}
monotone_map["price"] = -1 

# Defino entonces el vector de constraints en el orden esperado por XGB
monotone_constraints = tuple(monotone_map[name] for name in feature_names)

# Cheque que calze la longitud con la cantidad de features
assert len(monotone_constraints) == len(feature_names), "Dimensión de constraints no coincide con # de features."

# Ahora veo lo del Pipeline con restricciones monotónicas, o sea que con misma estructura de pasos pero pasándole el vector 
# de monotone_constraints al estimador

pipeline_xgb_wc = Pipeline(steps=[
    ("date_features", date_transformer),
    ("col_transformer", col_transformer),
    ("regresor_xgb", XGBRegressor(
        random_state=42,
        monotone_constraints=monotone_constraints,
        objective="reg:squarederror", 
        tree_method="hist", 
        n_jobs=-1
    ))
])

# Entrenamiento y predicción en validación
pipeline_xgb_wc.fit(X_train, y_train)
y_pred_val_xgb_wc = pipeline_xgb_wc.predict(X_val)

# Saco el MAE en validación (Punto 2)
mae_xgb_wc = MAE(y_val, y_pred_val_xgb_wc)
print("Valor MAE (XGB) Validación [wc = with constraints]:", f"{mae_xgb_wc:,.2f}")

# Ahora para poder guardar el modelo y las predicciones (Punto 4)
base_path = os.path.join(os.getcwd(), "Laboratorios", "Laboratorio_6", "data")
os.makedirs(base_path, exist_ok=True)

# lo de validación también lo guardo
pd.DataFrame({"y_val": y_val, "y_pred_xgb_wc": y_pred_val_xgb_wc}).to_csv(
    os.path.join(base_path, "predicciones_xgb_regressor_wc.csv"), index=False
)

# Guardar modelo este es con el pipe completo
with open(os.path.join(base_path, "xgb_regressor_model_wc.sav"), "wb") as f:
    pickle.dump(pipeline_xgb_wc, f)

# Cargar 
with open(os.path.join(base_path, "xgb_regressor_model_wc.sav"), "rb") as f:
    modelo_loaded_wc = pickle.load(f)
_ = modelo_loaded_wc.predict(X_val[:5])

Total features: 47
Columnas en orden del df: ['lat' 'long' 'pop' 'price' 'city_Athens' 'city_Irakleion' 'city_Larisa'
 'city_Patra' 'city_Thessaloniki' 'shop_shop_1' 'shop_shop_2'
 'shop_shop_3' 'shop_shop_4' 'shop_shop_5' 'shop_shop_6'
 'brand_adult-cola' 'brand_gazoza' 'brand_kinder-cola' 'brand_lemon-boost'
 'brand_orange-power' 'container_can' 'container_glass'
 'container_plastic' 'capacity_1.5lt' 'capacity_330ml' 'capacity_500ml'
 'day_28' 'day_29' 'day_30' 'day_31' 'month_1' 'month_2' 'month_3'
 'month_4' 'month_5' 'month_6' 'month_7' 'month_8' 'month_9' 'month_10'
 'month_11' 'month_12' 'year_2012' 'year_2013' 'year_2014' 'year_2015'
 'year_2016']
Valor MAE (XGB) Validación [wc = with constraints]: 2,903.11


**Respuesta desarrollo punto 3:** Al incorporar la restricción de monotonicidad negativa entre el precio y la cantidad, el error aumentó de ~2.800 a ~2900 unidades en MAE. Esto indica que el modelo con restricción es ligeramente menos preciso en términos de error absoluto promedio. Sin embargo, el resultado mantiene coherencia con la lógica económica, ya que al imponer que un aumento en el precio no pueda generar un incremento en la demanda, se reduce la flexibilidad del modelo y, en consecuencia, puede perder capacidad de ajuste en ciertos casos. Entonces, aunque el desempeño métrico empeora levemente, el modelo gana en interpretabilidad y consistencia teórica, confirmando la hipótesis del compañero ya que existe una relación inversa entre precio y cantidad demandada, pero imponerla estrictamente implica sacrificar una pequeña parte de la precisión predictiva en favor de un mayor realismo económico.

## 1.3 Optimización de Hiperparámetros con Optuna (20 puntos)

<p align="center">
  <img src="https://media.tenor.com/fmNdyGN4z5kAAAAi/hacking-lucy.gif">
</p>

Luego de presentarle sus resultados, Fiu le pregunta si es posible mejorar *aun más* su modelo. En particular, le comenta de la optimización de hiperparámetros con metodologías bayesianas a través del paquete `optuna`. Como usted es un aficionado al entrenamiento de modelos de ML, se propone implementar la descabellada idea de su jefe.

A partir de la mejor configuración obtenida en la sección anterior, utilice `optuna` para optimizar sus hiperparámetros. En particular, se pide que su optimización considere lo siguiente:

- Fijar una semilla en las instancias necesarias para garantizar la reproducibilidad de resultados
- Utilice `TPESampler` como método de muestreo
- De `XGBRegressor`, optimice los siguientes hiperparámetros:
    - `learning_rate` buscando valores flotantes en el rango (0.001, 0.1)
    - `n_estimators` buscando valores enteros en el rango (50, 1000)
    - `max_depth` buscando valores enteros en el rango (3, 10)
    - `max_leaves` buscando valores enteros en el rango (0, 100)
    - `min_child_weight` buscando valores enteros en el rango (1, 5)
    - `reg_alpha` buscando valores flotantes en el rango (0, 1)
    - `reg_lambda` buscando valores flotantes en el rango (0, 1)
- De `OneHotEncoder`, optimice el hiperparámetro `min_frequency` buscando el mejor valor flotante en el rango (0.0, 1.0)

Para ello se pide los siguientes pasos:
1. Implemente una función `objective()` que permita minimizar el `MAE` en el conjunto de validación. Use el método `.set_user_attr()` para almacenar el mejor pipeline entrenado. [10 puntos]
2. Fije el tiempo de entrenamiento a 5 minutos. [1 punto]
3. Optimizar el modelo y reportar el número de *trials*, el `MAE` y los mejores hiperparámetros encontrados. ¿Cómo cambian sus resultados con respecto a la sección anterior? ¿A qué se puede deber esto? [3 puntos] 
4. Explique cada hiperparámetro y su rol en el modelo. ¿Hacen sentido los rangos de optimización indicados? [5 puntos] 
5. Guardar su modelo en un archivo .pkl [1 punto]

In [6]:
import optuna
from optuna.samplers import TPESampler
optuna.logging.set_verbosity(optuna.logging.WARNING)

import time
import joblib

# Inserte su código acá

SEED = 42

# Espacios de búsqueda requeridos
def suggest_xgb_params(trial):
    # Fijamos grow_policy='lossguide' para que max_leaves tenga efecto (no se optimiza grow_policy)
    return {
        "learning_rate":   trial.suggest_float("learning_rate", 0.001, 0.1),
        "n_estimators":    trial.suggest_int("n_estimators", 50, 1000),
        "max_depth":       trial.suggest_int("max_depth", 3, 10),
        "max_leaves":      trial.suggest_int("max_leaves", 0, 100),
        "min_child_weight":trial.suggest_int("min_child_weight", 1, 5),
        "reg_alpha":       trial.suggest_float("reg_alpha", 0.0, 1.0),
        "reg_lambda":      trial.suggest_float("reg_lambda", 0.0, 1.0),
        # Fijos (no exigidos optimizar, pero correctos para regresión y performance)
        "objective":  "reg:squarederror",
        "tree_method":"hist",
        "grow_policy":"lossguide",
        "random_state": SEED,
        "n_jobs": -1,
    }

def suggest_ohe_params(trial):
    return {"min_frequency": trial.suggest_float("min_frequency", 0.0, 1.0)}

# Función objective  (Punto 1)
def objective(trial):
    params_xgb = suggest_xgb_params(trial)
    params_ohe = suggest_ohe_params(trial)

    # Pipelines unicos por trial (sin estado conocido desde antes por decir)
    date_tx  = clone(date_transformer)
    num_pipe = Pipeline([
        ("imp", SimpleImputer(strategy="median")),
        ("sc",  StandardScaler())
    ])
    cat_pipe = Pipeline([
        ("imp", SimpleImputer(strategy="most_frequent")),
        ("onehot", OneHotEncoder(handle_unknown="ignore",
                                 sparse_output=False,
                                 **params_ohe))
    ])
    col_tx = ColumnTransformer(
        [
            ("drop_ids", "drop", ["id"]),
            ("numerical",  num_pipe, ["lat", "long", "pop", "price"]),
            ("categorical",cat_pipe, ["city", "shop", "brand", "container",
                                      "capacity", "day", "month", "year"])
        ],
        verbose_feature_names_out=False,
        remainder="drop"
    ).set_output(transform="pandas")

    pipe = Pipeline(steps=[
        ("date_features", date_tx),
        ("col_transformer", col_tx),
        ("regresor_xgb", XGBRegressor(**params_xgb))
    ])

    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_val)
    mae = MAE(y_val, y_pred)

    # Guardar el pipeline de este trial en los atributos del trial
    trial.set_user_attr("pipeline", pipe)
    trial.set_user_attr("mae", float(mae))
    return mae

# Optimización a 5 minutos (Punto 2 y 3)
sampler = TPESampler(seed=SEED)
study = optuna.create_study(direction="minimize",
                            sampler=sampler,
                            study_name="opt_xgb_ohe_no_constraints")
study.optimize(objective, timeout=5*60, n_jobs=1)

# Reporte
print("\nNúmero de trials:", len(study.trials))
print("Mejor MAE:", study.best_value)
print("Mejores hiperparámetros:\n", study.best_params)

# Guardo mejor pipeline a .pkl (Punto 5)
best_pipe = study.best_trial.user_attrs["pipeline"]
out_dir = os.path.join(os.getcwd(), "Laboratorios", "Laboratorio_6", "data")
os.makedirs(out_dir, exist_ok=True)
joblib.dump(best_pipe, os.path.join(out_dir, "best_pipeline_optuna.pkl"))

y_pred_best = best_pipe.predict(X_val)
pd.DataFrame({"y_val": y_val, "y_pred_best": y_pred_best}).to_csv(
    os.path.join(out_dir, "predicciones_best_optuna.csv"),
    index=False
)
print("MAE (best, recomputado):", MAE(y_val, y_pred_best))

c:\ProgramData\anaconda3\envs\vis_info\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm



Número de trials: 133
Mejor MAE: 2466.01123046875
Mejores hiperparámetros:
 {'learning_rate': 0.07563366061598621, 'n_estimators': 877, 'max_depth': 10, 'max_leaves': 100, 'min_child_weight': 2, 'reg_alpha': 0.9035624887131728, 'reg_lambda': 0.30226186461351845, 'min_frequency': 0.04261335051877195}
MAE (best, recomputado): 2466.01123046875


**Respuesta desarrollo punto 3:** Luego de la optimización bayesiana con Optuna, el modelo alcanzó un MAE de ~2.460 en el conjunto de validación, tras 133 iteraciones. En comparación con el modelo sin optimización (MAE ≈ 2.678,35), se observa una mejora moderada en el desempeño, equivalente a una reducción de aproximadamente un ~13 % en el error absoluto promedio. Este resultado es coherente con el objetivo de la optimización, ya que el algoritmo TPE (Tree-structured Parzen Estimator) explora de manera eficiente el espacio de hiperparámetros, identificando combinaciones que equilibran la complejidad del modelo y su capacidad de generalización. En este caso, los mejores valores encontrados (por ejemplo, un learning_rate intermedio de ~0.07 y una mayor profundidad con max_depth = 10 y max_leaves = 100) permitieron capturar mejor las relaciones no lineales sin sobreajustar. Entonces, el ajuste de hiperparámetros mejoró la precisión del modelo, lo que confirma que una búsqueda sistemática y controlada de parámetros contribuye a un desempeño más robusto que el obtenido con los valores por defecto del algoritmo.

**Respuesta desarrollo punto 4:** Cada uno de los hiperparámetros optimizados cumple un rol específico en el comportamiento del modelo y su capacidad de generalización, ahora entenderemos un poco cómo es que influyen en el modelo y si los rangos definidos son adecuados:

- **learning_rate:** controla la magnitud de los pasos que da el modelo en cada iteración del proceso de boosting. Valores pequeños (como los explorados entre 0.001 y 0.1) permiten un aprendizaje más estable y gradual, reduciendo el riesgo de sobreajuste, aunque requieren un mayor número de iteraciones para converger. El rango utilizado es apropiado para garantizar un balance entre precisión y velocidad de aprendizaje.

- **n_estimators:** corresponde al número total de árboles que se construyen en el modelo. El rango entre 50 y 1000 es adecuado, ya que un número muy bajo puede producir subajuste, mientras que un número excesivo puede aumentar el riesgo de sobreajuste y el costo computacional.

- **max_depth:** define la profundidad máxima de los árboles individuales. El rango entre 3 y 10 permite explorar desde estructuras simples y generalistas hasta árboles más complejos que capturen interacciones no lineales sin sobreajustar.

- **max_leaves:** determina el número máximo de hojas permitidas por árbol cuando se usa la política de crecimiento *lossguide*. El rango entre 0 y 100 es lógico, ya que controla la granularidad de las divisiones pues valores bajos favorecen modelos más simples, mientras que valores altos incrementan la capacidad de ajuste.

- **min_child_weight:** establece el peso mínimo de las observaciones en un nodo hijo. Funciona como un regularizador que evita divisiones con pocas muestras, lo cual contribuye a controlar el sobreajuste. El rango (1–5) es adecuado, pues valores menores podrían generar árboles demasiado profundos y específicos, mientras que valores mayores podrían restringir en exceso el crecimiento del modelo.

- **reg_alpha:** representa la penalización L1 (Lasso) aplicada a los coeficientes del modelo. Ayuda a reducir la complejidad eliminando variables poco relevantes. El rango (0–1) permite ajustar su nivel de regularización de manera flexible.

- **reg_lambda:** es la penalización L2 (Ridge), que suaviza los coeficientes y ayuda a prevenir el sobreajuste sin forzar a que algunos parámetros se anulen completamente. El rango (0–1) también resulta razonable para controlar su intensidad.

- **min_frequency (OneHotEncoder):** especifica la frecuencia mínima con que debe aparecer una categoría para ser incluida como variable independiente en el proceso de codificación. Su rango (0–1) es coherente, ya que permite filtrar categorías poco frecuentes que podrían introducir ruido o provocar esparsidad excesiva en el conjunto de características.

Habíendo dicho todo esto entonces, podemos ver que los rangos de optimización definidos son adecuados y coherentes con las mejores prácticas de ajuste en modelos basados en gradient boosting, ya que permiten explorar un espacio suficientemente amplio para equilibrar precisión, interpretabilidad y capacidad de generalización. (Igual esto no asegura 100% que vamos a encontrar si o si el mejor modelo, pero si nos da una buena probabilidad de encontrar un buen modelo).

## 4. Optimización de Hiperparámetros con Optuna y Prunners (17 puntos)

<p align="center">
  <img src="https://i.pinimg.com/originals/90/16/f9/9016f919c2259f3d0e8fe465049638a7.gif">
</p>

Después de optimizar el rendimiento de su modelo varias veces, Fiu le pregunta si no es posible optimizar el entrenamiento del modelo en sí mismo. Después de leer un par de post de personas de dudosa reputación en la *deepweb*, usted llega a la conclusión que puede cumplir este objetivo mediante la implementación de **Prunning**.

Vuelva a optimizar los mismos hiperparámetros que la sección pasada, pero esta vez utilizando **Prunning** en la optimización. En particular, usted debe:

- Responder: ¿Qué es prunning? ¿De qué forma debería impactar en el entrenamiento? [2 puntos]
- Redefinir la función `objective()` utilizando `optuna.integration.XGBoostPruningCallback` como método de **Prunning** [10 puntos]
- Fijar nuevamente el tiempo de entrenamiento a 5 minutos [1 punto]
- Reportar el número de *trials*, el `MAE` y los mejores hiperparámetros encontrados. ¿Cómo cambian sus resultados con respecto a la sección anterior? ¿A qué se puede deber esto? [3 puntos]
- Guardar su modelo en un archivo .pkl [1 punto]

Nota: Si quieren silenciar los prints obtenidos en el prunning, pueden hacerlo mediante el siguiente comando:

```
optuna.logging.set_verbosity(optuna.logging.WARNING)
```

De implementar la opción anterior, pueden especificar `show_progress_bar = True` en el método `optimize` para *más sabor*.

Hint: Si quieren especificar parámetros del método .fit() del modelo a través del pipeline, pueden hacerlo por medio de la siguiente sintaxis: `pipeline.fit(stepmodelo__parametro = valor)`

Hint2: Este <a href = https://stackoverflow.com/questions/40329576/sklearn-pass-fit-parameters-to-xgboost-in-pipeline>enlace</a> les puede ser de ayuda en su implementación

In [7]:
# Inserte su código acá

from optuna.integration import XGBoostPruningCallback
optuna.logging.set_verbosity(optuna.logging.WARNING)
SEED = 42

# Repito el mismos espacios de búsqueda que la parte anterior
def suggest_xgb_params(trial):
    return {
        "learning_rate":    trial.suggest_float("learning_rate", 0.001, 0.1),
        "n_estimators":     trial.suggest_int("n_estimators", 50, 1000),
        "max_depth":        trial.suggest_int("max_depth", 3, 10),
        "max_leaves":       trial.suggest_int("max_leaves", 0, 100),
        "min_child_weight": trial.suggest_int("min_child_weight", 1, 5),
        "reg_alpha":        trial.suggest_float("reg_alpha", 0.0, 1.0),
        "reg_lambda":       trial.suggest_float("reg_lambda", 0.0, 1.0),
        # fijos coherentes con la otra parte
        "objective":  "reg:squarederror",
        "tree_method":"hist",
        "grow_policy":"lossguide",
        "random_state": SEED,
        "n_jobs": -1,
    }

def suggest_ohe_params(trial):
    return {"min_frequency": trial.suggest_float("min_frequency", 0.0, 1.0)}

# Redefino objective (Punto 2)

def objective(trial):
    params_xgb = suggest_xgb_params(trial)
    params_ohe = suggest_ohe_params(trial)
    # Pipelines únicos por trial (sin estado conocido desde antes por decir)
    date_tx  = clone(date_transformer)
    num_pipe = Pipeline([("imp", SimpleImputer(strategy="median")),
                         ("sc",  StandardScaler())])
    cat_pipe = Pipeline([("imp", SimpleImputer(strategy="most_frequent")),
                         ("onehot", OneHotEncoder(handle_unknown="ignore",
                                                  sparse_output=False,
                                                  **params_ohe))])
    col_tx = ColumnTransformer(
        [("drop_ids", "drop", ["id"]),
         ("numerical",  num_pipe, ["lat", "long", "pop", "price"]),
         ("categorical",cat_pipe, ["city", "shop", "brand", "container",
                                   "capacity", "day", "month", "year"])],
        verbose_feature_names_out=False,
        remainder="drop"
    ).set_output(transform="pandas")

    # De ahí transformar datos
    pipe_pre = Pipeline([("date_features", date_tx),
                         ("col_transformer", col_tx)])
    X_train_trans = pipe_pre.fit_transform(X_train, y_train)
    X_val_trans   = pipe_pre.transform(X_val)

    # definimos el prunnignng callback
    pruning_cb = XGBoostPruningCallback(trial, "validation_1-mae")

    # Mmodelo xgb con pruning (fit directo, fuera del Pipeline)
    model = XGBRegressor(
        **params_xgb,
        eval_metric="mae",
        callbacks=[pruning_cb]
    )
    model.fit(
        X_train_trans, y_train,
        eval_set=[(X_train_trans, y_train), (X_val_trans, y_val)],
        verbose=False
    )

    # Evaluar MAE en validación
    y_pred = model.predict(X_val_trans)
    mae = MAE(y_val, y_pred)

    # Guardar pipeline completo en atributos del trial
    full_pipe = Pipeline([("date_features", date_tx),
                          ("col_transformer", col_tx),
                          ("regresor_xgb", model)])
    trial.set_user_attr("pipeline", full_pipe)
    trial.set_user_attr("mae", float(mae))
    return mae

# Limito a 5 minutos (Punto 3)
sampler = TPESampler(seed=SEED)
study = optuna.create_study(direction="minimize",
                            sampler=sampler,
                            study_name="opt_xgb_ohe_pruning")
study.optimize(objective, timeout=5*60, n_jobs=1, show_progress_bar=True)

# Reporte requerido (Punto 4)
print("\nNúmero de trials:", len(study.trials))
print("Mejor MAE:", study.best_value)
print("Mejores hiperparámetros:\n", study.best_params)

# Guardo mejor modelo .pkl (Punto 5)
best_pipe = study.best_trial.user_attrs["pipeline"]
out_dir = os.path.join(os.getcwd(), "Laboratorios", "Laboratorio_6", "data")
os.makedirs(out_dir, exist_ok=True)
joblib.dump(best_pipe, os.path.join(out_dir, "best_pipeline_optuna_pruning.pkl"))

Best trial: 21. Best value: 2571.74:  100%|██████████| 05:08/05:00



Número de trials: 65
Mejor MAE: 2571.74462890625
Mejores hiperparámetros:
 {'learning_rate': 0.08218065461343788, 'n_estimators': 875, 'max_depth': 6, 'max_leaves': 56, 'min_child_weight': 3, 'reg_alpha': 0.7882937674472152, 'reg_lambda': 0.21781112379201073, 'min_frequency': 0.06934727309538372}


['c:\\Users\\56966\\OneDrive - Universidad de Chile\\Escritorio\\Laboratorio de programación\\TamoJosha-Lab-de-Prog-MDS-Prim2025\\Laboratorios\\Laboratorio_6\\Laboratorios\\Laboratorio_6\\data\\best_pipeline_optuna_pruning.pkl']

**Respuesta desarrollo punto 1:** El pruning es una técnica utilizada durante la optimización de modelos para detener anticipadamente aquellos trials que muestran bajo desempeño parcial en las primeras etapas del entrenamiento. En lugar de completar todas las iteraciones de cada configuración de hiperparámetros, el algoritmo evalúa periódicamente su rendimiento y descarta de forma temprana los modelos que no prometen mejorar el mejor resultado actual. En el contexto de Optuna, esta técnica se implementa mediante el callback `XGBoostPruningCallback`, el cual monitorea una métrica de validación (en este caso, el `MAE`) durante el proceso de entrenamiento. Si el desempeño del modelo se mantiene significativamente por debajo del mejor valor registrado, el trial es interrumpido, liberando recursos computacionales para explorar nuevas combinaciones más prometedoras. En simple, el pruning reduce el tiempo total de búsqueda y mejora la eficiencia del proceso de optimización, ya que concentra el esfuerzo de cómputo en configuraciones con mayor potencial. Sin embargo, su aplicación también puede generar ligeras variaciones en el número de trials ejecutados o incluso en el valor final del MAE, dependiendo de los criterios de corte, dado que algunos modelos podrían detenerse antes de alcanzar su mejor desempeño posible.

**Respuesta desarrollo punto 4:** Al incorporar el pruning dentro del proceso de optimización con Optuna, se obtuvieron 65 trials completados, con un MAE de ~2.570 en el conjunto de validación. 

Los mejores hiperparámetros encontrados fueron:`learning_rate = 0.082`, `n_estimators = 875`, `max_depth = 6`, `max_leaves = 56`, `min_child_weight = 3`, `reg_alpha = 0.79`, `reg_lambda = 0.22` y `min_frequency = 0.07`.

En comparación con la sección anterior (sin pruning), el número de trials disminuyó y el valor del MAE aumentó levemente. Esta diferencia es esperable, ya que el pruning interrumpe de forma anticipada aquellos trials que muestran un desempeño poco prometedor, reduciendo así el tiempo total de optimización pero a costa de explorar menos a fondo ciertas configuraciones. Esto implica un compromiso entre eficiencia computacional y precisión pues el pruning acelera la búsqueda al descartar combinaciones subóptimas, pero también puede omitir modelos que habrían mejorado con más iteraciones. Aun así, el resultado se mantiene dentro del mismo orden de magnitud que la optimización sin pruning, lo cual demuestra que el proceso fue estable y consistente, validando la efectividad del ajuste automático con control de recursos.


## 5. Visualizaciones (5 puntos)

<p align="center">
  <img src="https://media.tenor.com/F-LgB1xTebEAAAAd/look-at-this-graph-nickelback.gif">
</p>


Satisfecho con su trabajo, Fiu le pregunta si es posible generar visualizaciones que permitan entender el entrenamiento de su modelo.

A partir del siguiente <a href = https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/005_visualization.html#visualization>enlace</a>, genere las siguientes visualizaciones:

1. Gráfico de historial de optimización [1 punto]
2. Gráfico de coordenadas paralelas [1 punto]
3. Gráfico de importancia de hiperparámetros [1 punto]

Comente sus resultados:

4. ¿Desde qué *trial* se empiezan a observar mejoras notables en sus resultados? [0.5 puntos]
5. ¿Qué tendencias puede observar a partir del gráfico de coordenadas paralelas? [1 punto]
6. ¿Cuáles son los hiperparámetros con mayor importancia para la optimización de su modelo? [0.5 puntos]

In [8]:
# Inserte su código acá
from optuna.visualization import (
    plot_optimization_history,
    plot_parallel_coordinate,
    plot_param_importances,
)

# Gráfico de historial de optimización (Punto 1)
fig_hist = plot_optimization_history(study)
fig_hist.show()

# Gráfico de coordenadas paralelas (Punto 2)
fig_pc = plot_parallel_coordinate(
    study,
    params=[
        "learning_rate", "n_estimators", "max_depth", "max_leaves",
        "min_child_weight", "reg_alpha", "reg_lambda", "min_frequency"
    ],
)
fig_pc.show()

# Gráfico de importancia de hiperparámetros (Punto 3)
fig_imp = plot_param_importances(study)
fig_imp.show()


**Respuesta desarrollo punto 4:** A partir del gráfico de historial de optimización se observa que las mejoras más significativas en el valor objetivo ocurren durante los primeros trials, particularmente entre el trial 0 y el trial 20, donde el error (MAE) disminuye drásticamente desde valores superiores a 9.000 hasta estabilizarse cerca de los 2.800. Después de ese punto, las variaciones se vuelven marginales, indicando que el algoritmo de búsqueda encontró una región óptima del espacio de hiperparámetros y que las mejoras posteriores fueron menores.

**Respuesta desarrollo punto 5:** En el gráfico de coordenadas paralelas se aprecian ciertas tendencias en los hiperparámetros asociados a mejores resultados. Los trials con menor error (tonos más oscuros) tienden a presentar un learning_rate medio-alto (entre 0.05 y 0.1), una profundidad max_depth intermedia (cercana a 6), y valores de max_leaves moderados (alrededor de 60–70). Asimismo, los modelos con mejor desempeño muestran min_frequency baja (inferior a 0.1), lo cual sugiere que incluir categorías poco frecuentes aportó información relevante para el ajuste del modelo. Estas relaciones reflejan un equilibrio entre complejidad del modelo y capacidad de generalización.

**Respuesta desarrollo punto 6:** Según el gráfico de importancia de hiperparámetros, los factores que más influyeron en la optimización del modelo fueron reg_alpha (penalización L1), min_frequency (frecuencia mínima del OneHotEncoder) y learning_rate. Esto indica que la regularización y la representación categórica fueron determinantes para reducir el error, mientras que otros parámetros como max_depth, reg_lambda y n_estimators tuvieron un impacto menor. Los resultados sugieren que la capacidad del modelo para evitar el sobreajuste (a través de la regularización y la codificación eficiente) fue más relevante que aumentar su complejidad estructural.

## 6. Síntesis de resultados (3 puntos)

Finalmente:

1. Genere una tabla resumen del MAE en el conjunto de validación obtenido en los 5 modelos entrenados desde Baseline hasta XGBoost con Constraints, Optuna y Prunning. [1 punto]
2. Compare los resultados de la tabla y responda, ¿qué modelo obtiene el mejor rendimiento? [0.5 puntos]
3. Cargue el mejor modelo, prediga sobre el conjunto de **test** y reporte su MAE. [0.5 puntos]
4. ¿Existen diferencias con respecto a las métricas obtenidas en el conjunto de validación? ¿Porqué puede ocurrir esto? [1 punto]

In [9]:
# Inserte su código acá
from sklearn.metrics import mean_absolute_error as MAE
from IPython.display import display

# Carpeta donde están modelos
base_path = os.path.join("Laboratorios", "Laboratorio_6", "data")
os.makedirs(base_path, exist_ok=True)

# Rescato los modelos y calculo MAE en validación
def _load_model(path):
    if path.endswith(".pkl") and "best_pipeline_optuna" in path:
        return joblib.load(path)
    elif path.endswith(".pkl") and ("dummy_regressor_model" in path or "xgb_regressor_model" in path):
        with open(path, "rb") as f:
            return pickle.load(f)
    elif path.endswith(".sav"):
        with open(path, "rb") as f:
            return pickle.load(f)
    else:
        with open(path, "rb") as f:
            return pickle.load(f)

def _mae_en_valid(modelo, X_val, y_val):
    y_pred = modelo.predict(X_val)
    return MAE(y_val, y_pred)

# Rutas de los 5 modelos
rutas_modelos = {
    "DummyRegressor (baseline)": os.path.join(base_path, "dummy_regressor_model.pkl"),
    "XGBRegressor (default)":    os.path.join(base_path, "xgb_regressor_model.pkl"),
    "XGBRegressor (constraints)":os.path.join(base_path, "xgb_regressor_model_wc.sav"),
    "XGB + Optuna":              os.path.join(base_path, "best_pipeline_optuna.pkl"),
    "XGB + Optuna + Pruning":    os.path.join(base_path, "best_pipeline_optuna_pruning.pkl"),
}

# Genero tabla MAE (Punto 1)
filas = []
for nombre, ruta in rutas_modelos.items():
    try:
        modelo = _load_model(ruta)
        mae_val = _mae_en_valid(modelo, X_val, y_val)
        filas.append({"modelo": nombre, "ruta": ruta, "MAE_valid": mae_val})
    except Exception as e:
        filas.append({"modelo": nombre, "ruta": ruta, "MAE_valid": None})
        print(f"[WARN] No se pudo evaluar {nombre} en validación. Error: {e}")

df_mae_val = (
    pd.DataFrame(filas)
    .sort_values("MAE_valid", na_position="last")
    .reset_index(drop=True)
)

print("\nResumen MAE (validación)")
display(df_mae_val)
df_mae_val.to_csv(
    os.path.join(base_path, "resumen_mae_validacion.csv"),
    index=False
)

# Cargo mejor modelo (Punto 3)
mejor_fila = df_mae_val.iloc[0]
mejor_modelo_nombre = mejor_fila["modelo"]
mejor_modelo_ruta   = mejor_fila["ruta"]

print(f"\nMejor modelo según validación: {mejor_modelo_nombre}")
print(f"Ruta: {mejor_modelo_ruta}")

mejor_modelo = _load_model(mejor_modelo_ruta)
y_pred_test = mejor_modelo.predict(X_test)
mae_test = MAE(y_test, y_pred_test)

print(f"\nMAE (TEST) del mejor modelo [{mejor_modelo_nombre}]: {mae_test:,.4f}")

pd.DataFrame({"y_test": y_test, "y_pred_test": y_pred_test}).to_csv(
    os.path.join(base_path, "predicciones_mejor_modelo_test.csv"), index=False
)


Resumen MAE (validación)


modelo  \
0                XGB + Optuna   
1      XGB + Optuna + Pruning   
2      XGBRegressor (default)   
3  XGBRegressor (constraints)   
4   DummyRegressor (baseline)   

                                                ruta     MAE_valid  
0  Laboratorios\Laboratorio_6\data\best_pipeline_...   2466.011230  
1  Laboratorios\Laboratorio_6\data\best_pipeline_...   2571.744629  
2  Laboratorios\Laboratorio_6\data\xgb_regressor_...   2796.359131  
3  Laboratorios\Laboratorio_6\data\xgb_regressor_...   2903.112061  
4  Laboratorios\Laboratorio_6\data\dummy_regresso...  12518.387191


Mejor modelo según validación: XGB + Optuna
Ruta: Laboratorios\Laboratorio_6\data\best_pipeline_optuna.pkl

MAE (TEST) del mejor modelo [XGB + Optuna]: 3,366.2449


**Respuesta desarrollo parte 2:** El modelo con mejor rendimiento en validación es **XGB + Optuna**, que alcanza un MAE de ~2.470, superando al resto de las configuraciones. Este resultado refleja que la optimización de hiperparámetros mediante Optuna y permitió ajustar de forma precisa la complejidad del modelo, mejorando su capacidad predictiva respecto a las versiones anteriores (más efectiva al equilibrar complejidad, regularización y poder predictivo.).
En comparación, el modelo **XGB + Optuna + Pruning** obtiene un MAE levemente superior (~2.570), lo que indica que el pruning no mejoró el desempeño, posiblemente porque detuvo anticipadamente algunas configuraciones prometedoras.
El modelo **XGBRegressor (default)** logra un MAE de ~2.800, mostrando un avance importante frente al **DummyRegressor (12.518,39)**, pero sin alcanzar el nivel de ajuste del modelo optimizado.
Por último, la versión con restricciones monotónicas (MAE = ~2900) mantiene coherencia teórica, aunque con una ligera pérdida de precisión.

**Respuesta desarrollo punto 4:** El mejor modelo en validación fue **XGB + Optuna**, con un MAE de ~2.470, mientras que en el conjunto de test el error aumentó a ~3.370. Esta diferencia es esperable y se debe principalmente a que el modelo fue ajustado y evaluado durante la optimización utilizando únicamente los datos de entrenamiento y validación. El conjunto de test representa información completamente nueva que el modelo no ha visto antes, por lo tanto, el incremento del error refleja la capacidad real de generalización del modelo. En este caso, el aumento moderado del MAE sugiere que el modelo mantiene un buen desempeño fuera de la muestra usada para calibrarlo, aunque es inevitable cierta pérdida de precisión. Entonces, la diferencia entre ambos conjuntos se explica por la variabilidad inherente de los datos, el riesgo de sobreajuste parcial durante la optimización y el hecho de que la validación tiende a ser más "optimista", ya que influye directamente en el proceso de selección de hiperparámetros.

### Nota: Se ponen valores aproximados ya que al correr localmente en distintos equipos, cifras cambian ligeramente, pero se mantiene la interpretación.

# Conclusión
Exito!
<p align="center">
  <img src="https://i.pinimg.com/originals/55/3d/42/553d42bea9b10e0662a05aa8726fc7f4.gif">
</p>